compiler/rustc_mir_transform/src/const_prop_lint.rs

```diff
@@ -286,25 +285,13 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
     }
 
     /// Returns the value, if any, of evaluating `c`.
-    fn eval_constant(
-        &mut self,
-        c: &Constant<'tcx>,
-        _source_info: SourceInfo,
-    ) -> Option<OpTy<'tcx>> {
+    fn eval_constant(&mut self, c: &Constant<'tcx>, source_info: SourceInfo) -> Option<OpTy<'tcx>> {
         // FIXME we need to revisit this for #67176
         if c.needs_subst() {
             return None;
         }
 
-        match self.ecx.const_to_op(&c.literal, None) {
-            Ok(op) => Some(op),
-            Err(error) => {
-                let tcx = self.ecx.tcx.at(c.span);
-                let err = ConstEvalErr::new(&self.ecx, error, Some(c.span));
-                err.report_as_error(tcx, "erroneous constant used");
```

Comment:
maybe we can get the best of both worlds by using

```rust
tcx.span_err(c.span, "erroneous constant used");
```

potentially along with some check avoiding the error if `c.literal` is `Unevaluated` and the def id's span overlaps with `c.span`. That should get rid of the duplication

Reply:
I don't think this should even be an `err`. It's just a note/help pointing out where the other error originated from. I see we have `span_note_without_error`, maybe that would make sense?

Reply:
yea that seems reasonable.

Reply:
Why does the error in const-err-late even look so different from e.g. defaults-not-assumed-fail? The code looks very similar, but in one case there is a CTFE error at the `<() as Tr>::B` and in the other case there is no such error at the `S::<i32>::FOO`.

Reply:
Oh, I think in one case it got promoted so there is an intermediate constant -- there are two CTFE errors, one from the inner constant and one from the outer where it references the inner.

So we are quite inconsistent in how we handle faulty constants between when they occur in const context vs in runtime context. Not sure which version I prefer.